### PR TITLE
Refactor `NewMessageForm` component

### LINKED ISSUE #7493: Simplify new message form
Group the functionality in the new message form's attachments menu into a smaller child component.

### LINKED ISSUE #9040: Special symbols replaced when use EmojiPicker
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Steps to reproduce
1. Type on of the `<`, `>`
2. Pick an emoji using the EmojiPicker
3. See result

### Expected behaviour
Tell us what should happen

### Actual behaviour

https://user-images.githubusercontent.com/93392545/224549080-c8a99812-cd07-41e6-831f-64d3162810cd.mp4

Should be the issue with NcRichContenteditable, which parses all the symbols when paste something. Doesn't affect the resulted message, though

## Talk app

**Talk app version:** (16 + 17)

### LINKED ISSUE #9511: Adjust layout of the template picker with server update
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

* Follow https://github.com/nextcloud/server/pull/38174

### PR DESCRIPTION
**Pull request by Antreesy**
### ☑️ Resolves

* Fix #7493 
* Fix #9040 
* Fix #9511 

### 🚧 Tasks

- Splitted by commits for easier review :blue_heart: 
- [x] Rename existing components according to Vue styleguide
- [x] Minor renamings of properties, typo and lint fixes
- [x] Extract attachments NcActions to separate component
- [x] Extract text creation NcModal to separate component
- [ ] Visual check
- [ ] Code review

### 🏁 Checklist

- [x] ⛑️ Tests (unit and/or integration) are included or not required
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [x] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required
- [x] 🔖 Capability is added or not needed

### CODE REVIEW

**Comment by Antreesy** [comment]
* Rebased onto master
* Move recently added composable for viewer to child component
---
### To discuss:
Shall we rename a parent component to make all names shorter? `NewMessageForm` => `NewMessage`

**Comment by Antreesy** [comment]
* Rebased onto master (with TypingIndicator)
* Add a commit that resolves #9511

File: src/components/NewMessageForm/NewMessageFormAttachments.vue [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+		:disabled="disabled"
+		:aria-label="t('spreed', 'Share files to the conversation')"
+		:aria-haspopup="true">
+		<template #icon>
+			<Paperclip :size="16" />
+		</template>
+
+		<NcActionButton v-if="canUploadFiles"
+			close-after-click
+			@click.prevent="$emit('open-file-upload')">
```
Comment (nickvergessen) [inline src/components/NewMessageForm/NewMessageFormAttachments.vue:37]: The "open-file-upload" event has been triggered but not declared on `emits` option

File: src/components/NewMessageForm/NewMessageForm.vue [hunk 2]
```diff
@@ -641,13 +498,13 @@ export default {
 		async handleSubmitSpam(numberOfMessages) {
 			console.debug('Sending ' + numberOfMessages + ' lorem ipsum messages')
 			for (let i = 0; i < numberOfMessages; i++) {
-				const randomNumber = parseInt(Math.random() * 500, 10)
+				const randomNumber = parseInt((Math.random() * 500).toString(), 10)
```
Comment (ShGKme) [inline src/components/NewMessageForm/NewMessageForm.vue:501]:
```suggestion
				const randomNumber = Math.floor(Math.random() * 500)
```

Or tricky but short 😀

```suggestion
				const randomNumber = ~~(Math.random() * 500)
```

Comment (ShGKme) [inline src/components/NewMessageForm/NewMessageForm.vue:501]:
```suggestion
				const randomNumber = Math.floor(Math.random() * 500)
```

Or tricky but short 😀

```suggestion
				const randomNumber = ~~(Math.random() * 500)
```

File: src/components/NewMessageForm/NewMessageFormTextCreateDialog.vue [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+			required: true,
+		},
+
+		showTextFileDialog: {
+			type: Number,
+			required: true,
+		},
+	},
+
+	emits: ['dismiss-text-file-creation'],
```
Comment (ShGKme) [inline src/components/NewMessageForm/NewMessageFormTextCreateDialog.vue:110]:
```suggestion
	emits: ['dismiss'],
```

This component is `TextCreateDialog`, so we may not mention its name in the event name.

File: src/components/NewMessageForm/NewMessageFormTextCreateDialog.vue [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+				'--border': border + 'px',
+				'--fullwidth': width + 2 * (margin + border) + 'px',
+				'--height': this.fileTemplate.ratio ? Math.round(width / this.fileTemplate.ratio) + 'px' : null,
+			}
+		},
+	},
+
+	mounted() {
+		const fileTemplate = this.fileTemplateOptions[this.showTextFileDialog]
+		this.textFileTitle = fileTemplate.label + fileTemplate.extension
```
Comment (ShGKme) [inline src/components/NewMessageForm/NewMessageFormTextCreateDialog.vue:169]: Why should it be after the mounting?

File: src/components/NewMessageForm/NewMessageFormPollEditor.vue [hunk 5]
```diff
@@ -184,27 +192,30 @@ export default {
...
-:deep(.checkbox-radio-switch__label) {
-	align-items: unset;
-	height: unset;
-	margin: 4px 0;
-	padding: 8px;
-	width: 100%;
-	border-radius: var(--border-radius-large);
-	:deep(span) {
-		align-self: flex-start;
+	// Upstream
```
Comment (ShGKme) [inline src/components/NewMessageForm/NewMessageFormPollEditor.vue:214]:
```suggestion
	// FIXME Upstream
```

File: src/components/NewMessageForm/NewMessageFormPollEditor.vue [hunk 6]
```diff
@@ -24,83 +25,90 @@
...
+			<div v-for="(option, index) in pollOptions"
 				:key="index"
-				:ref="`pollOption${index}`"
-				class="simple-polls-editor__option"
-				:value.sync="pollOptions[index]"
-				:label="t('spreed', 'Answer {option}', {option: index + 1})"
-				:can-delete="pollOptions.length > 2"
-				@delete-option="deleteOption(index)" />
+				class="poll-editor__option">
+				<NcTextField :ref="`pollOption${index}`"
```
Comment (ShGKme) [inline src/components/NewMessageForm/NewMessageFormPollEditor.vue:44]:
```suggestion
				<NcTextField ref="pollOption"
```

Both Vue 2 and Vue 3 support using ref with `v-for`, the result is an array.

https://v2.vuejs.org/v2/guide/components-edge-cases.html#Accessing-Child-Component-Instances-amp-Child-Elements

https://vuejs.org/guide/essentials/template-refs.html#refs-inside-v-for

It is a bit more clean (just an array, no creating strings with index inside) and easier to use with Composition API.

File: src/components/NewMessageForm/NewMessageForm.vue [hunk 7]
```diff
@@ -169,93 +139,51 @@
...
+		<NewMessageFormPollEditor v-if="showPollEditor"
 			:token="token"
-			@close="toggleSimplePollsEditor(false)" />
+			@close="togglePollEditor" />
 
 		<!-- Text file creation dialog -->
-		<NcModal v-if="showTextFileDialog !== false"
-			size="normal"
+		<NewMessageFormTextCreateDialog v-if="showTextFileDialog !== -1"
+			ref="textFileCreateDialog"
```
Comment (ShGKme) [inline src/components/NewMessageForm/NewMessageForm.vue:152]: Seems, this ref is unused

File: src/components/NewMessageForm/NewMessageFormTextCreateDialog.vue [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+import NcModal from '@nextcloud/vue/dist/Components/NcModal.js'
+import NcTextField from '@nextcloud/vue/dist/Components/NcTextField.js'
+
+import NewMessageFormTemplatePreview from './NewMessageFormTemplatePreview.vue'
+
+import { useViewer } from '../../composables/useViewer.js'
+import { createTextFile, shareFile } from '../../services/filesSharingServices.js'
+
+export default {
+	name: 'NewMessageFormTextCreateDialog',
```
Comment (ShGKme) [inline src/components/NewMessageForm/NewMessageFormTextCreateDialog.vue:84]: Not sure about the name "TextCreate". For general "text" creation we use the main input field. But this dialog is for creating a file.

I'd call something like:
- `NewTextFileDialog`
- `TextFileDialog`
- `ShareNewTextFileDialog`

Also, it seems that it is a pure component, not coupled with the new message form, except for opening the viewer. We may (but do not have to) make it a shared component, not `NewMessageForm_Part`.

**Review by ShGKme — COMMENTED** [review]
Mostly looks good to me, makes this super component simpler and smaller.
I added some not blocking comments.

Tested, that all still work in main chat view and breakout rooms.

**Comment by ShGKme** [comment]
> Shall we rename a parent component to make all names shorter?

I agree to rename it. It's shorter, the component is actually more, than a `form` (includes attachments, and a voice message), and its interface is not similar to the form interface (no `submit` event or `reset` method etc).

File: src/components/NewMessageForm/NewMessageFormAttachments.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (Antreesy) [reply src/components/NewMessageForm/NewMessageFormAttachments.vue:37]: We have a lot of NcButtons with `@click.prevent`, should be a follow-up

File: src/components/NewMessageForm/NewMessageFormTextCreateDialog.vue (hunk 4 above)
Comment (Antreesy) [reply src/components/NewMessageForm/NewMessageFormTextCreateDialog.vue:169]: Moved to watch()

File: src/components/NewMessageForm/NewMessageFormTextCreateDialog.vue (hunk 8 above)
Comment (Antreesy) [reply src/components/NewMessageForm/NewMessageFormTextCreateDialog.vue:84]: As long as it is used only in one place, I'd like to keep the name
Also, this is used not for .md files, but also for other extensions, so renamed it into `NewMessageNewFileDialog`

**Comment by Antreesy** [comment]
Rebased onto master because of conflict with https://github.com/nextcloud/spreed/pull/9526
Commits after review are here: https://github.com/nextcloud/spreed/pull/9430/commits/5ae4c8207f03887ecfd6fec5f6510a1d66689888, https://github.com/nextcloud/spreed/pull/9430/commits/1ffd8e58910e048a07fabc123834cb1d5b7ed075

**Comment by ShGKme** [comment]
I disabled auto-merge, as it is a bit large refactoring in case @marcoambrosini would want to review it too.

Feel free to re-enable if you want it to be in the beta release.

**Review by ShGKme — APPROVED** [review]
Still works after changes and rebase 😀

**Pull request merged**